### ISSUE TITLE
Need help adding a new type

### ISSUE THREAD

**Issue by gzz2000** [ealeksandrov/ProvisionQL#50]
Hi Evgeny, I tried to enhance ProvisionQL by adding support to an alias of the .ipa file, .tipa (TrollStore IPA, an airdrop-friendly extension making it easy to send ipa files to iOS devices running the [TrollStore](https://github.com/opa334/TrollStore) semi-jailbreak, which is very popular these days).

I did so by modifying all code related to ipa file handling and add tipa to the matching list (see [my commit](https://github.com/gzz2000/ProvisionQL/commit/3294c2f0874bea494980d67051fcf3f8997aa931)). By `mdls -name kMDItemContentType` I found the type should be `com.opa334.trollstore.tipa`. I've succeeded in making it working if I explicitly launch my modified ProvisionQL like this:

```
qlmanage -p file.tipa -g ~/Library/QuickLook/ProvisionQL.qlgenerator -c com.opa334.trollstore.tipa
```

But it does not work if I do not explicitly specify the `-g` and `-c` options, nor in Finder directly.
Do you have any idea why it doesn't work? Did I miss something to edit? Thanks!

**Comment by relikd** [comment]
tell me what this command returns for your tipa file:
```sh
mdls -name kMDItemContentType your_app.tipa
```

**Comment by relikd** [comment]
Interesting, so there is a macOS application for tipa files. I did not find any on the official homepage.
With this command you can check which App is used:

```sh
/System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Versions/A/Support/lsregister -dump Type | grep '^uti:\s*com.opa334.trollstore.tipa$' -B2 -A3 | grep '^flags:\s*active' -B5 
```

you tried changing it in your source code or using my branch?
I've pushed a new version, try again.

**Comment by relikd** [comment]
Oh, and my opinion on that is that it is probably better to create a separate QL plugin for that. The TIPA file does not follow the same structure as normal IPAs. In fact, it would be best if the app you have would add a QL extension to its code.

The one tipa I tried has the icon file stored in a .car asset bundle. ProvisionQL (currently) does not support extracting from .car files. At least Apple provided .ipa files will always bundle an Artwork which can be used as icon. There are simply no guarantees as to what will and will not work in tipa.

**Comment by relikd** [comment]
Does .tipa work with this plugin? (remember to `qlmanage -r`)
[ProvisionQL.qlgenerator.zip](https://github.com/ealeksandrov/ProvisionQL/files/14158411/ProvisionQL.qlgenerator.zip)

**Comment by gzz2000** [comment]
This is working! I have tested a few (~10) .tipa files and most of them works with this plugin.
However, 2 of them did not seem to work on my side. I don't know if it's because of caches on my side or something special about these two files. I uploaded the failed ones for reference.
[failed_tipa.zip](https://github.com/ealeksandrov/ProvisionQL/files/14159704/failed_tipa.zip)

**Comment by relikd** [comment]
Yes, the error happens because of a bug in the original code (tried to access non-existing key `CFBundleShortVersionString`). I fixed that in my overall refactoring. Here is the newest version which builds atop my refactoring branch with `.tipa` enabled.
[ProvisionQL_with_tipa.zip](https://github.com/ealeksandrov/ProvisionQL/files/14167011/ProvisionQL_with_tipa.zip)

Note: TrollDecrypt does not show an icon but this is due to the poor quality control of tipa developers. In this case, the developer created an empty icon file with @3x which is prefered over the @1x resolution. As mentioned above, I noticed other inconsistencies which may and will break the functionality of this plugin. But at least you get as much information as possible out of it.

**Comment by gzz2000** [comment]
This newest version works on all my .tipa files. This is cool!

**Comment by relikd** [comment]
You shouldn't close the issue, as it is not added to the main branch yet. waiting for @ealeksandrov's approval

**Comment by gzz2000** [comment]
Reopened:)

**Comment by ealeksandrov** [comment]
@gzz2000 thanks for sharing the info about `tipa`. Looks like TrollStore [doesn't support](https://github.com/opa334/TrollStore) iOS 17.0.1+
Is it still useful today? Is it valuable to work on `tipa` preview? Maybe something new is already out? :)

**Comment by gzz2000** [comment]
> [@gzz2000](https://github.com/gzz2000) thanks for sharing the info about `tipa`. Looks like TrollStore [doesn't support](https://github.com/opa334/TrollStore) iOS 17.0.1+ Is it still useful today? Is it valuable to work on `tipa` preview? Maybe something new is already out? :)

I believe TrollStore still has lots of users today. And it's one of the very few cases where an end user has an ipa file on their Mac.

**Comment by relikd** [comment]
Hey @gzz2000, I dont know if `.tipa` will ever be added to `ProvisionQL`, as my branch probably wont be merged. Anyway, I've started my own QuickLook plugin [relikd/QLAppBundle](https://github.com/relikd/QLAppBundle) which already has `.tipa` support and runs on 10.15 - 15.7 (havent tested 26 yet). enjoy.
